src/UI/HTMLTooltip.js

```diff
@@ -110,17 +145,50 @@ export class HTMLTooltip {
         this.left = left
         this.top = top
 
-        if (this.transformRuleIndex && shadow.styleSheets[0].rules[this.transformRuleIndex]) {
-            // If we have already set the rule, remove it…
-            shadow.styleSheets[0].deleteRule(this.transformRuleIndex)
+        const ruleObj = this.transformRules[element]
+
+        if (ruleObj.index) {
+            if (shadow.styleSheets[0].rules[ruleObj.index]) {
+                // If we have already set the rule, remove it…
+                shadow.styleSheets[0].deleteRule(ruleObj.index)
+            }
         } else {
             // …otherwise, set the index as the very last rule
-            this.transformRuleIndex = shadow.styleSheets[0].rules.length
+            ruleObj.index = shadow.styleSheets[0].rules.length
         }
 
-        let cssRule = this.options.tooltipPositionClass?.(top, left)
+        const cssRule = ruleObj.getRuleString?.(top, left)
         if (typeof cssRule === 'string') {
-            shadow.styleSheets[0].insertRule(cssRule, this.transformRuleIndex)
+            shadow.styleSheets[0].insertRule(cssRule, ruleObj.index)
+        }
+
+        if (this.options.hasCaret) {
+            const tooltipBoundingBox = this.tooltip.getBoundingClientRect()
+            // If overflowing from the left, try centering it in the window
+            if (tooltipBoundingBox.left < 0) {
+                const leftPosWhenCentered = (window.innerWidth - tooltipBoundingBox.width) / 2
+                const overriddenLeftPosition = left + Math.abs(tooltipBoundingBox.left) + leftPosWhenCentered
+
+                this.updatePosition(element, {left: overriddenLeftPosition, top})
+            }
+
+            // If overflowing from the right, move it slightly to the left
+            if (tooltipBoundingBox.right > window.innerWidth) {
+                const rightOverflow = window.innerWidth - tooltipBoundingBox.right
+                const overriddenLeftPosition = left - Math.abs(rightOverflow) - 5
+
+                this.updatePosition(element, {left: overriddenLeftPosition, top})
+            }
```

Comment:
I tried several combinations of these and they work pretty decently. It's certainly an improvement over the previous state. It never gets into an infinite loop because we [constrain the width to the viewport](https://github.com/duckduckgo/duckduckgo-autofill/blob/main/src/UI/styles/autofill-tooltip-styles.css#L29) so it can never overflow both ends. I tried removing the rule and the test fails so we're covered.